### PR TITLE
Better processor groups in P matrix construction [pmatrix-groups-fix]

### PR DESCRIPTION
**Pull request by jakubcerveny**
This is a permanent fix following PR #503:
* Shared vertices, edges and faces in ParNCMesh are what they need to be for DG and visualization (as in slave-faces-fix)
* But the P matrix communication groups are now constructed independently of that and should be optimal (so that slave DOFs are not getting sent anywhere)
* The code should actually be simpler and more readable now. The P matrix groups are constructed explicitly with `ParNCMesh::CalculatePMatrixGroups()` and the suspicious method `ParNCMesh::AugmentMasterGrooups()` is gone, along with its ugly state variable.

I did some testing on random meshes. I also ran the AMR benchmark on vulcan on 64K cores. It runs OK and we got back those ~10% of performance.

### CODE REVIEW

**Comment by tzanio** [comment]
Did you see the Travis issues (MPI, no DEBUG)?

File: mesh/pncmesh.cpp [hunk 1]
```diff
@@ -249,30 +262,37 @@ void ParNCMesh::ElementSharesEdge(int elem, int enode)
    int &owner = tmp_owner[e_index];
    owner = std::min(owner, el_rank);
 
-   index_rank.Append(Connection(e_index, el_rank));
+   char &flag = tmp_shared_flag[e_index];
+   if (el_rank == MyRank) { flag |= 0x1; }
+   if (el_rank != MyRank) { flag |= 0x2; }
```
Comment (v-dobrev) [inline mesh/pncmesh.cpp]: This second `if` can be replaced with `else`.

File: mesh/pncmesh.cpp [hunk 2]
```diff
@@ -285,7 +305,11 @@ void ParNCMesh::ElementSharesVertex(int elem, int vnode)
    int &owner = tmp_owner[v_index];
    owner = std::min(owner, el_rank);
 
-   index_rank.Append(Connection(v_index, el_rank));
+   char &flag = tmp_shared_flag[v_index];
+   if (el_rank == MyRank) { flag |= 0x1; }
+   if (el_rank != MyRank) { flag |= 0x2; }
```
Comment (v-dobrev) [inline mesh/pncmesh.cpp]: Same here: `if (...)` -> `else`.

**Comment by jakubcerveny** [comment]
Do we want to mention this in the CHANGELOG or is it too small?

**Comment by v-dobrev** [comment]
Regarding the CHANGELOG - it's up to you - I'm fine either way. After that, if you want, you can merge this branch into `next` for testing.

**Comment by tzanio** [comment]
I'll mention it briefly in the CHANGELOG.

**Comment by tzanio** [comment]
@jakubcerveny can you make final changes (if any), merge in `next` and add the `in-next` label?

**Pull request merged**